File: examples/java/java-booking-02_introduce_parameter_object/src/main/java/tv/codely/booking/Booking.java
package tv.codely.booking;

import java.time.LocalDateTime;

public final class Booking {
    private final BookingId     id;
    private final LocalDateTime startDate;
    private final LocalDateTime endDate;
    private final CustomerId    customerId;
    private final CustomerName  customerName;
    private final EmailAddress  customerEmail;
    private final BookingType   bookingType;
    private final DiscountType  discountType;
    private final DiscountValue discountValue;
    private final TaxType       taxType;
    private final TaxValue      taxValue;

    public Booking(
        BookingId id,
        DateRange bookingDateRange,
        Customer customer,
        BookingType bookingType,
        Discount discount,
        Tax tax
    ) {
        this.id            = id;
        this.startDate     = bookingDateRange.startDate();
        this.endDate       = bookingDateRange.endDate();
        this.customerId    = customer.id();
        this.customerName  = customer.name();
        this.customerEmail = customer.emailAddress();
        this.bookingType   = bookingType;
        this.discountType  = discount.type();
        this.discountValue = discount.value();
        this.taxType       = tax.type();
        this.taxValue      = tax.value();
    }

    public BookingStatus statusFor(LocalDateTime date) {
        if (date.isBefore(startDate)) {
            return BookingStatus.NOT_STARTED;
        }

        if (date.isAfter(startDate) && date.isBefore(endDate)) {
            return BookingStatus.ACTIVE;
        }

        return BookingStatus.FINISHED;
    }
}

File: examples/csharp/csharp-booking-02_introduce_parameter_object/src/Booking/Booking.cs
using System;

namespace CodelyTv.Booking
{
    public sealed class Booking
    {
        private readonly BookingId id;
        private readonly DateTime startDate;
        private readonly DateTime endDate;
        private readonly CustomerId customerId;
        private readonly CustomerName customerName;
        private readonly EmailAddress customerEmail;
        private readonly BookingType bookingType;
        private readonly DiscountType discountType;
        private readonly DiscountValue discountValue;
        private readonly TaxType taxType;
        private readonly TaxValue taxValue;

        public Booking(
            BookingId id,
            DateRange bookingDateRange,
            Customer customer,
            BookingType bookingType,
            Discount discount,
            Tax tax
        )
        {
            this.id = id;
            this.startDate = bookingDateRange.StartDate;
            this.endDate = bookingDateRange.EndDate;
            this.customerId = customer.Id;
            this.customerName = customer.Name;
            this.customerEmail = customer.EmailAddress;
            this.bookingType = bookingType;
            this.discountType = discount.Type;
            this.discountValue = discount.Value;
            this.taxType = tax.Type;
            this.taxValue = tax.Value;
        }

        public BookingStatus StatusFor(DateTime date)
        {
            if (date < startDate)
            {
                return BookingStatus.NOT_STARTED;
            }

            if (date > startDate && date < endDate)
            {
                return BookingStatus.ACTIVE;
            }

            return BookingStatus.FINISHED;
        }
    }
}
